### PR TITLE
TTD -- misc fixes and enhancements.

### PR DESCRIPTION
**Pull request by mrkmarron**
- Add support for additional promise object.
- Disable code cache for TTD.
- Fix handling of names on snap/restore external function.
- Add missing well known functions to core enumeration.
- Revise script context reuse heuristics in re-inflate. Check for problematic objects and enable more re-use.

### CODE REVIEW

File: lib/Runtime/Debug/TTSnapObjects.cpp [hunk 1]
```diff
@@ -153,23 +147,73 @@ namespace TTD
...
+                    {
                         propertyReset.Clear();
-                        return dynObj->GetLibrary()->CreateObject();
+                        return true;
                     }
+                }
+            }
+
+            propertyReset.Clear();
+            ////
```
Comment (kfarnung) [inline lib/Runtime/Debug/TTSnapObjects.cpp]: Unneeded slashes?

File: lib/Runtime/Debug/TTSnapObjects.cpp [hunk 2]
```diff
@@ -153,23 +147,73 @@ namespace TTD
...
+        }
+
+        Js::DynamicObject* ObjectPropertyReset_WellKnown(const SnapObject* snpObject, Js::DynamicObject* dynObj, InflateMap* inflator)
+        {
+            TTDAssert(snpObject->OptWellKnownToken != TTD_INVALID_WELLKNOWN_TOKEN, "Should only call this on well known objects.");
+
+            JsUtil::BaseHashSet<Js::PropertyId, HeapAllocator>& propertyReset = inflator->GetPropertyResetSet();
+            propertyReset.Clear();
+
+            ////
```
Comment (kfarnung) [inline lib/Runtime/Debug/TTSnapObjects.cpp]: nit: Here too

File: lib/Runtime/Library/GlobalObject.cpp [hunk 3]
```diff
@@ -617,7 +617,15 @@ namespace Js
                 Throw::FatalInternalError();
             }
 #endif
+
+#if ENABLE_TTD
+            if(!scriptContext->IsTTDRecordOrReplayModeEnabled())
+            {
+                scriptContext->AddToEvalMap(key, isIndirect, pfuncScript);
```
Comment (digitalinfinity) [inline lib/Runtime/Library/GlobalObject.cpp]: What is the reason for this change?

**Comment by digitalinfinity** [comment]
:shipit:

File: lib/Runtime/Library/GlobalObject.cpp (hunk 3 above)
Comment (mrkmarron) [inline lib/Runtime/Library/GlobalObject.cpp]: We disabled the cache during record (since trying to replay the behavior was problematic) so no need to reset cache values here.
